File: node/src/invoices.ts
import { LinopayConfigError } from './errors.js';
import type { HttpClient } from './http.js';
import type { ChannelTokenCache } from './tokens.js';
import type { LinopayConfig } from './config.js';

/**
 * Flexi-Payments (invoicing), pay-now scope only.
 *
 * The SOW (§1, Flexi-Payment bullet) is explicit: do NOT build an SDK
 * method for the "customer commits to a future date" capability the
 * company messaging describes — that backend work isn't on `main` yet.
 * Scope is exactly:
 *   - create a pay-now invoice (dormant QR)
 *   - get one (re-read, so the merchant screen can settle)
 *   - cancel / recall
 *
 * Auth: every call in this group uses the exchanged scoped token
 * (NOT the raw channel-signed JWT). The exchanged token carries
 * `invoices:write` for create/cancel, `invoices:read` for get — but
 * `invoices:write` implies read access too (see ChannelAccessToken.
 * HasScope), so a single cached token suffices for all three.
 */
export interface CreateInvoiceInput {
  amountCents: number;
  currency?: string;
  reference?: string;
  paymentWindowDays?: number;
}

export interface Invoice {
  invoiceId: string;
  invoiceCode: string;
  amount: number;
  currency: string;
  reference: string | null;
  status: string;
  dueAt: string;
  expiresAt: string;
  paymentWindowDays: number;
  qrCodeUrl: string;
  qrCodeImage: string;
}

export interface InvoicesClient {
  create(input: CreateInvoiceInput): Promise<Invoice>;
  get(invoiceId: string): Promise<Invoice>;
  cancel(invoiceId: string): Promise<Invoice & { cancelledAt?: string }>;
}

export function createInvoicesClient(
  _cfg: LinopayConfig,
  http: HttpClient,
  tokens: ChannelTokenCache,
): InvoicesClient {
  void _cfg; // reserved for per-call explicit key-id / merchant-id overrides
  return {
    async create(input: CreateInvoiceInput): Promise<Invoice> {
      const amountCents = input.amountCents;
      if (!Number.isInteger(amountCents) || amountCents <= 0) {
        throw new LinopayConfigError(
          `amountCents must be a positive integer (got ${JSON.stringify(amountCents)}).`,
        );
      }
      const token = await tokens.exchange(['invoices:write']);
      const body = {
        amount: Number((amountCents / 100).toFixed(2)),
        currency: input.currency ?? 'NZD',
        reference: input.reference ?? null,
        paymentWindowDays: input.paymentWindowDays ?? 14,
      };
      type Response = Invoice;
      const resp = await http.callJson<Response>(
        `/api/merchants/${encodeURIComponent(token.merchantId)}` +
          `/channels/${encodeURIComponent(token.channelId)}/invoices`,
        {
          method: 'POST',
          token: token.accessToken,
          body,
        },
      );
      return resp;
    },

    async get(invoiceId: string): Promise<Invoice> {
      if (!invoiceId) {
        throw new LinopayConfigError('invoiceId is required to look up an invoice.');
      }
      const token = await tokens.exchange(['invoices:read']);
      type Response = Invoice;
      const resp = await http.callJson<Response>(
        `/api/merchants/${encodeURIComponent(token.merchantId)}/invoices/${encodeURIComponent(invoiceId)}`,
        { method: 'GET', token: token.accessToken },
      );
      return resp;
    },

    async cancel(invoiceId: string): Promise<Invoice & { cancelledAt?: string }> {
      if (!invoiceId) {
        throw new LinopayConfigError('invoiceId is required to cancel an invoice.');
      }
      const token = await tokens.exchange(['invoices:write']);
      type Response = Invoice & { cancelledAt?: string };
      const resp = await http.callJson<Response>(
        `/api/merchants/${encodeURIComponent(token.merchantId)}/invoices/${encodeURIComponent(invoiceId)}/cancel`,
        { method: 'POST', token: token.accessToken, body: {} },
      );
      return resp;
    },
  };
}

File: dotnet/src/LinoPay.Sdk/Invoices.cs
using System.Text.Json.Serialization;

namespace Linopay.Sdk;

/// <summary>
/// Flexi-Payments (invoicing), pay-now scope only.
///
/// <para>
/// The SOW (§1, Flexi-Payment bullet) is explicit: do NOT build an
/// SDK method for the "customer commits to a future date" capability
/// the company messaging describes — that backend work isn't on
/// <c>main</c> yet. Scope is exactly:
///   - create a pay-now invoice (dormant QR)
///   - get one (re-read, so the merchant screen can settle)
///   - cancel / recall
/// </para>
///
/// <para>
/// Auth: every call in this group uses the exchanged scoped token
/// (NOT a raw channel-signed JWT). The exchanged token carries
/// <c>invoices:write</c> for create/cancel, <c>invoices:read</c> for
/// get — but <c>invoices:write</c> implies read access too (see
/// <c>ChannelAccessToken.HasScope</c>), so a single cached token
/// suffices for all three.
/// </para>
/// </summary>
public sealed class InvoicesClient
{
    private readonly Config _config;
    private readonly LinopayHttp _http;
    private readonly ChannelTokenCache _tokens;

    public InvoicesClient(Config config, LinopayHttp http, ChannelTokenCache tokens)
    {
        _config = config.Validate();
        _http = http;
        _tokens = tokens;
    }

    public async Task<Invoice> CreateAsync(CreateInvoiceArgs args, CancellationToken ct = default)
    {
        if (args.AmountCents <= 0)
        {
            throw new LinopayConfigException(
                $"AmountCents must be a positive integer (got {args.AmountCents}).");
        }

        var token = await _tokens.ExchangeAsync(new[] { "invoices:write" }, ct);
        var body = new
        {
            amount = (args.AmountCents / 100m).ToString("F2", System.Globalization.CultureInfo.InvariantCulture),
            currency = args.Currency ?? "NZD",
            reference = args.Reference ?? null,
            paymentWindowDays = args.PaymentWindowDays ?? 14,
        };
        var resp = await _http.CallAsync<Invoice>(
            $"/api/merchants/{Uri.EscapeDataString(token.MerchantId)}" +
            $"/channels/{Uri.EscapeDataString(token.ChannelId)}/invoices",
            new CallOptions { Method = "POST", Token = token.AccessToken, Body = body },
            ct);
        return resp;
    }

    public async Task<Invoice> GetAsync(string invoiceId, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(invoiceId))
        {
            throw new LinopayConfigException("InvoiceId is required to look up an invoice.");
        }

        var token = await _tokens.ExchangeAsync(new[] { "invoices:read" }, ct);
        var resp = await _http.CallAsync<Invoice>(
            $"/api/merchants/{Uri.EscapeDataString(token.MerchantId)}/invoices/{Uri.EscapeDataString(invoiceId)}",
            new CallOptions { Method = "GET", Token = token.AccessToken },
            ct);
        return resp;
    }

    public async Task<Invoice> CancelAsync(string invoiceId, CancellationToken ct = default)
    {
        if (string.IsNullOrWhiteSpace(invoiceId))
        {
            throw new LinopayConfigException("InvoiceId is required to cancel an invoice.");
        }

        var token = await _tokens.ExchangeAsync(new[] { "invoices:write" }, ct);
        var resp = await _http.CallAsync<Invoice>(
            $"/api/merchants/{Uri.EscapeDataString(token.MerchantId)}/invoices/{Uri.EscapeDataString(invoiceId)}/cancel",
            new CallOptions { Method = "POST", Token = token.AccessToken, Body = new { } },
            ct);
        return resp;
    }
}

public sealed record CreateInvoiceArgs(
    int AmountCents,
    string? Currency = null,
    string? Reference = null,
    int? PaymentWindowDays = null);

public sealed record Invoice
{
    [JsonPropertyName("invoiceId")]           public string InvoiceId { get; set; } = "";
    [JsonPropertyName("invoiceCode")]         public string InvoiceCode { get; set; } = "";
    [JsonPropertyName("amount")]              public decimal Amount { get; set; }
    [JsonPropertyName("currency")]            public string Currency { get; set; } = "";
    [JsonPropertyName("reference")]           public string? Reference { get; set; }
    [JsonPropertyName("status")]              public string Status { get; set; } = "";
    [JsonPropertyName("dueAt")]               public string DueAt { get; set; } = "";
    [JsonPropertyName("expiresAt")]           public string ExpiresAt { get; set; } = "";
    [JsonPropertyName("paymentWindowDays")]   public int PaymentWindowDays { get; set; }
    [JsonPropertyName("qrCodeUrl")]           public string QrCodeUrl { get; set; } = "";
    [JsonPropertyName("qrCodeImage")]         public string QrCodeImage { get; set; } = "";
}
